beetsplug/substitute.py

```diff
@@ -46,11 +48,32 @@ def __init__(self):
         Get the configuration, register template function and create list of
         substitute rules.
         """
-        super().__init__()
+        super(Substitute, self).__init__()
         self.substitute_rules = []
         self.template_funcs["substitute"] = self.tmpl_substitute
 
-        for key, view in self.config.items():
-            value = view.as_str()
-            pattern = re.compile(key.lower())
+        if isinstance(self.config.get(), dict):
+            self._log.warning(
+                "Unordered configuration is deprecated, as it leads to"
+                + " unpredictable behaviour on overlapping rules.\n"
+                + textwrap.dedent(
+                    """
+                    Old syntax:
+                      substitute:
+                        a: b
+                        c: d
+
+                    New syntax:
+                      substitute:
+                      - a: b
+                      - c: d
```

Comment:
✍🏼 Nit:

```suggestion
                        - a: b
                        - c: d
```

Reply:
I believe this is somewhat of a divisive matter :)

I too used the indented syntax you are suggesting for many years, but after repeatedly encountering both varieties being used (often even within the same document). I eventually got tired of the inconsistency and tried to find out which was "correct" (both are of course valid yaml).

It seems the [yaml specification](https://yaml.org/spec/1.2.2/#example-mapping-scalars-to-sequences) favours the less indented variety and uses it in all its examples. If beets favours the indented variety, I am of course happy to be consistent.

Reply:
Regarding the codebase, I thought the plugin documentation may be suitable place to check:

```fish
$ grep '^ *- [a-z]+: ' -iEB2 docs/**/*.rst
docs/plugins/advancedrewrite.rst-
docs/plugins/advancedrewrite.rst-    advancedrewrite:
docs/plugins/advancedrewrite.rst:      - match: "mb_artistid:dec0f331-cb08-4c8e-9c9f-aeb1f0f6d88c year:..2022"
--
docs/plugins/advancedrewrite.rst-
docs/plugins/advancedrewrite.rst-    advancedrewrite:
docs/plugins/advancedrewrite.rst:      - match: "artist:배유빈 feat. 김미현"
--
docs/plugins/advancedrewrite.rst-    advancedrewrite:
docs/plugins/advancedrewrite.rst-      # BAD, DON'T DO THIS!
docs/plugins/advancedrewrite.rst:      - match: album:THE ALBUM
--
docs/plugins/advancedrewrite.rst-    advancedrewrite:
docs/plugins/advancedrewrite.rst-      # Note the quotes around the query string!
docs/plugins/advancedrewrite.rst:      - match: album:"THE ALBUM"
--
docs/plugins/albumtypes.rst-    albumtypes:
docs/plugins/albumtypes.rst-        types:
docs/plugins/albumtypes.rst:            - ep: 'EP'
docs/plugins/albumtypes.rst:            - single: 'Single'
docs/plugins/albumtypes.rst:            - soundtrack: 'OST'
docs/plugins/albumtypes.rst:            - live: 'Live'
docs/plugins/albumtypes.rst:            - compilation: 'Anthology'
docs/plugins/albumtypes.rst:            - remix: 'Remix'
--
docs/plugins/fetchart.rst-        sources:
docs/plugins/fetchart.rst-            - filesystem
docs/plugins/fetchart.rst:            - coverart: release
docs/plugins/fetchart.rst-            - itunes
docs/plugins/fetchart.rst:            - coverart: releasegroup
```

It seems like the indentation is preferred.

I personally just follow whatever is the preference of the most popular formatter for the language: in this case both `yamlfix` and `yamlfmt` add indentation by default.

Reply:
I was going to make the change, but I guess it's not needed now, as the list format won't be needed :)